### ISSUE TITLE
Slow Range Join

### ISSUE THREAD

**Issue by max-hoffman** [dolthub/dolt#6298]
# Background

Range joins are join queries with inequality ON conditions:
```sql
tmp3> create table xy (x int primary key, y int);
tmp3> create table uv (u int primary key, v int);
tmp3> explain select * from uv join xy on x <= u and u < y;
+----------------------------------------+
| plan                                   |
+----------------------------------------+
| InnerJoin                              |
|  ├─ ((xy.x <= uv.u) AND (uv.u < xy.y)) |
|  ├─ Table                              |
|  │   ├─ name: xy                       |
|  │   └─ columns: [x y]                 |
|  └─ Table                              |
|      ├─ name: uv                       |
|      └─ columns: [u v]                 |
+----------------------------------------+
```

We will focus on single conjuction expressions that are prefixes of sorted indexes.

We currently perform these as NESTED_LOOP_JOIN because we do not have a way to execute them as lookup, hash, or merge joins.

For big tables, nested loop joins will be slow (nxm complexity). Sometimes that is unavoidable, for example if all LHS rows match all RHS rows under the range. But most of the time we will only match a fraction of the RHS rows.

Specialized range join operators might outperform NESTED_LOOP_JOIN. For example, we could implement a LOOKUP_JOIN for range expressions. LOOKUP_JOIN is normally only applied to equijoins that are point lookups and have appropriate indexes for lookups. A range LOOKUP_JOIN would also require indexes, but use an indexes range access to only read a subset of rows. If  the LHS expression matches the prefix a sorted index, we could also use a two pointer algorithm to perform a sliding window (only read the RHS table once). In some cases it might be beneficial to perform these operations in-memory, rather than going to storage for repeated calls.

# Implementation

Basically we have a customer with an extremely slow range join that matches the conditions above. The LHS table is 900K rows, the RHS is 8 rows. The standalone query matches 0 rows, but still takes ~2 minutes. This range query is part of a larger 13-table join, and in the larger join it has a larger performance impact.

The priority here is finding something that makes the customer workload faster. That will require
- Benchmarks that at least replicate the customer's data shape (900k row LHS, 8 row RHS, 0 rows match condition)
- Fast prototype of range LOOKUP_JOIN and range SLIDING_JOIN
- Rethink strategy if neither of those is fast
- Follow-up with polished implementation and correctness tests so that we can release and test the impact on the larger join query

**Comment by max-hoffman** [comment]
Basically we have a customer with an extremely slow range join that matches the conditions above. The LHS table is 900K rows, the RHS is 8 rows. The standalone query matches 0 rows, but still takes ~2 minutes. This range query is part of a larger 13-table join, and in the larger join it has a larger performance impact.

**Comment by nicktobey** [comment]
Update: This will be fixed as soon as Dolt gets updated to use the most recent version of go-mysql-server.

For context: Max's suggestion worked like a charm, and we also found ways to adjust the coster to choose a better plan for another slow join in the same query. The query is now twice as fast.

**Comment by nicktobey** [comment]
Fixed by https://github.com/dolthub/go-mysql-server/pull/1880 and https://github.com/dolthub/go-mysql-server/pull/1889
